### DISCUSSION TITLE
Licence de publication pour Gipsy v2

### DISCUSSION THREAD

**Discussion by Aeris1One** [Gunivers/Gipsy#33] [Ideas]
Actuellement, et malgré le code sur GitHub, Gipsy v1 n'est **pas open-source** car il n'a été publié sous aucune licence. Il est donc en théorie interdit de réutiliser son code sans demander l'autorisation individuelle de **tous les contributeurs** du projet.
À titre d'information, la majorité des autres projets de Gunivers sont sous licence Creative Commons.

La fondation Creative Commons elle-même recommande de ne pas utiliser ses licences pour des logiciels car elles n'y sont pas adaptées.

Il faudrait donc chercher ailleurs, par exemple autour de la GNU GPL, la licence la plus répandue du logiciel libre, ou ses dérivées, l'AGPL qui demanderait à quiconque utiliserait de façon publique le bot modifié de republier les changements sous la même licence pour qu'ils soient éventuellement intégrés au bot principal.
Étant donné qu'il est compliqué d'utiliser le bot comme une librairie, la licence MIT imposerait sensiblement les mêmes limites que l'AGPL.

Le gros problème de ces licences est qu'elles sont adaptées au cadre juridique américain et non pas au cadre européen, ni français.
C'est pour pallier à ça que je propose d'utiliser une licence conçue pour l'Europe, comme l'EUPL (European Union Public Licence) ou la CéCILL.
À noter cependant : l'EUPL n'est pas compatible avec la GPL, ce qui signifie qu'il n'est pas possible d'utiliser des bibliothèques sous GPL. De plus, l'EUPL permettant à la personne qui possède une copie du logiciel la possibilité de le redistribuer sous une autre licence, beaucoup ayant un copyleft plus faible, et donc possibilité à terme de rendre un fork plus ou moins propriétaire.

La CéCILL me semble la meilleure, la GNU AGPL en second, pour Gipsy. En effet elle permettrait à n'importe qui de pouvoir réutiliser et repartager le code de façon légale. Avec cependant une contrainte : si une personne modifie le bot et le redistribue (soit en republiant une version self-hostable du code, soit en le rendant "accessible via un réseau" autrement dit en le connectant à Discord) elle doit nécessairement republier le code modifié sous CéCILL ou AGPL selon la licence choisie.

À noter que cela ne s'applique pas aux plugins, puisqu'il ne s'agit pas d'une version modifiée du bot. Les plugins, s'ils ne sont pas intégrés directement dans le code du bot, peuvent être publiés par leurs auteurs sous n'importe quelle licence, étant donné que c'est l'utilisateur final qui compose lui-même son bot.

Dernière chose: pour changer la licence du bot il est nécessaire que tous les détenteurs de droits sur le code actuel soient d'accord unanimement sur la licence à appliquer. Un seul refus empêche la publication. 
Il est possible de changer la licence après coup, mais les personnes ayant téléchargé le code avant un changement de licence, ayant "signé" le contrat illimité de cette licence, pourront continuer à l'utiliser. En clair, si l'on change de licence, cela signifie arrêter de proposer aux gens de signer un contrat (l'ancienne licence) et proposer un autre contrat (la nouvelle), les personnes en question pourront soit ne pas signer le nouveau contrat mais ne pourrons plus utiliser les nouveaux codes disponible uniquement sous la nouvelle licence, soit signer la nouvelle licence et se conformer à celle-ci.

De plus, il est possible de licencier plusieurs composants du bot sous des licences compatibles (c'est à dire deux contrats impossible à satisfaire en même temps). À la discrétion de celui qui possède le code du composant en question. Même si c'est évidemment peu recommandé pour ne pas avoir à gérer une infinité de licences dans le même projet. Il est possible de définir une condition pour n'accepter que les PR sous la même licence que le bot principal celà dit.

Qu'en pensez-vous ?

**Comment by Aeris1One** [comment]
Pour clarifier, une licence c'est principalement un contrat du style : "je t'autorise à utiliser mon code à condition que..." entre le propriétaire d'un morceau de code et l'utilisateur.

La loi stipule qu'il est nécessaire de demander la permission de quelqu'un quand on veut utiliser quelque-chose qu'il possède, et le contrat de licence est donc une autorisation sous conditions à tout le monde d'utiliser le code qu'on possède.


Chacun est libre de choisir sous quelle licence il publie ses codes, voire sous aucune licence, dans ce cas il faut demander la permissions pour utiliser, même si le code est disponible publiquement.
 
Certaines licences sont dites "copyright" ("c'est à moi vous n'utilisez pas", typiquement les EULAs de logiciels propriétaires, comme Minecraft, dont le code est accessible mais interdit d'utiliser)

D'autres sont "permissives" ("vous faites ce que vous voulez", parfois avec une obligation de créditer l'auteur original : Licence WTF, MIT, Unlicence)

D'autres sont dites "copyleft" ("vous pouvez utiliser tant que vous aussi vous publiez votre truc sous une licence copyleft" : (A/L)GPL, CéCILL)



Voilà pour l'explication rapide au cas où, pour plus d'informations le site de la GNU est franchement top, bien qu'assez partial et prônant ouvertement le copyleft et la GNU GPL : https://www.gnu.org/licenses/licenses.html

**Comment by VForiel** [comment]
Honnêtement je suis un peu perdu... *y'a beaucoup de trucs à lire* 😅

En soit pour Gunivers, le plus souvent on a besoin d'une licence qui:
- Autorise la réutilisation en open-source et crédité avec la même licence
- Autorise l'utilisation public si crédité
- Autorise l'utilisation privée
- Interdit l'utilisation commerciale
- Interdit par défaut la redistribution (on doit donner notre accord explicitement)

**Reply by Aeris1One** [reply]
1️⃣ La GNU GPL:
- autorise la redistribution tant que c'est sous licence GPL également et crédité
- n'interdit pas explicitement la redistribution commerciale (elle oblige tout de même à placer cette redistribution sous GPL, donc à publier le code source)
- par "redistribution" elle entends "partager le code source ou un binaire modifié ou non"
Elle n'est pas adaptée à un bot Discord car elle ne s'appliquerait pas si l'utilisateur ne faisait qu'héberger un bot dérivé sans partager le code source ou un binaire.

2️⃣ La GNU AGPL:
- même conditions que la GPL
- par "redistribution" elle entends "mettre à disposition le logiciel soit par republication du code source ou d'un binaire, soit par un accès réseau à celui-ci"
La clause d'"accès réseau" protégerait un bot Discord, c'est-à-dire quelqu'un qui modifierait Gipsy et le connecterais à Discord devrait republier le code sous AGPL.

Pour résumer la GPL en une phrase : "tu as le droit de repartager (même de façon payante)/modifier/faire un produit dérivé de mon travail tant que tu laisses les crédits et que toi aussi tu publies sous cette licence"

Il est à noter que la GPL ne définit pas de juridiction compétente pour l'appliquer, qu'elle est en stricte anglais et que plusieurs clauses sont illégales en France.



3️⃣ La CeCILL:
- autorise la redistribution tant que c'est sous la même licence
- n'interdit pas l'exploitation commerciale, tant que le code résultant de cette exploitation est sous CeCILL également (donc ouvert)
- elle inclut l'accès via réseau au terme de  "redistribution"
- elle précise que, sauf si le propriétaire du code et l'utilisateur se mettent d'accord, la juridiction française est compétente
- elle est compatible avec la loi française
- elle est disponible en anglais et en français


Je n'ai pas trop compris ton dernier point sur la redistribution, "redistribuer" signifie pour la plupart de ces licences soit partager le code source (faire un fork), soit partager le binaire (python c'est compliqué), soit le rendre accessible via le réseau (le mettre en ligne, seulement AGPL et CeCILL).


*Nota: J'ai considéré que "utilisation publique" et "utilisation privée" voulaient dire "bot public" et "bot privé" au sens de Discord. Légalement c'est la même chose, puisque dans tous les cas, à moins d'être seul sur le serveur avec le bot, c'est de la redistribution*

**Reply by VForiel** [reply]
Par redistribution j'entendais surtout le fait de proposer le bot en téléchargement via des groupes/sites web tiers

**Reply by VForiel** [reply]
Ce qu'il faudrait éviter

**Reply by Aeris1One** [reply]
C'est justement le principe du logiciel libre...
Tu auras aucune licence qui dit "vous devez forcément utiliser GitHub" ou "vous ne pouvez pas redistribuer le code non modifié" parce que ça serait contre l'idéologie du libre.
Mais ouais dans l'idéal faudrait que les gens passent par notre GitHub, c'est pour ça qu'il faudrait mettre un header à chaque fichier du bot rappelant la licence, ainsi que le lien du GitHub. Après bon, y'a pas vraiment de stats sur GitHub donc c'est pas comme une map sur Curseforge où on évite la redistribution...

**Reply by VForiel** [reply]
Ouai mais y'a des stats sur nos liens de DL, docn même si les téléchargements directs depuis Github ne sont pas comptabilisés, le fait de passer par nous aide quand même. Sans compter le fait qu'on mette des liens avec une étape de soutient pour inciter les utilisateurs à contribuer à Gunivers. Si on autorise la redistribution sans contrôle, n'importe qui pourrait profiter de la communauté ... et c'est déjà un peu le cas pour pas mal de sites à l'éthique douteuse.

Que les créations de la commu soient open-source c'est une chose, mais qu'elle soit libre au sens ou n'importe qui peut se l'approprier, le redistribuer et/ou se faire de l'argent dessus, personellement ça m'enchante pas tellement :/

**Reply by Aeris1One** [reply]
À vrai dire c'est là-dessus que se base l'opensource, tout le monde peut redistribuer une version améliorée.
Les personnes mal intentionnées qui veulent simplement se faire de l'argent sur les créations de la communauté vont à priori retirer les crédits, ne pas faire accepter la licence au téléchargement, ou nous refuser d'obtenir le code source par exemple, dans tous les cas il s'agit d'une violation de la licence, qui appelle donc une action en justice et retire les droits d'utilisation dans de tels cas (même si c'est chiant d'appeler la justice, ça permet au moins le coup de pression "arrêtez sinon tribunal")

Si elles redistribuent en gardant les crédits et en apportant des modifications, en respectant tous les termes de la licence, il y a de grandes chances que ces personnes soient de bonne foi. On pourrait donc sans doute s'arranger avec elles pour passer directement par notre site, quitte à faire de cette personne un contributeur (c'est toujours jouissif quand une de ses modifications non-officielle est intégrée dans le projet original)

Et puis de toute façon il y aura des comportements douteux, licence ou non. 
Si nous mettons une licence limitante, les personnes mal intentionnées s'en branleront royalement et continueront quand-même au mépris de celle-ci, et les personnes honnêtes seront limitées dans leurs possibilités.
